### PR TITLE
digest, registry/storage, registry/handlers: switch to SHA256 as canonical digest

### PR DESCRIPTION
**Pull request by endophage**
Supersedes https://github.com/docker/distribution/pull/237

@jlhawn @stevvooe

### CODE REVIEW

File: digest/digest.go [hunk 1]
```diff
@@ -16,6 +16,8 @@ import (
 const (
 	// DigestTarSumV1EmptyTar is the digest for the empty tar file.
 	DigestTarSumV1EmptyTar = "tarsum.v1+sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+	// DigestSha256EmptyTar is the canonical sha256 digest of empty data
+	DigestSha256EmptyTar = "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
```
Comment (diogomonica) [inline digest/digest.go]: This makes me so happy.

File: registry/storage/layerupload.go [hunk 2]
```diff
@@ -11,7 +11,6 @@ import (
 	ctxu "github.com/docker/distribution/context"
 	"github.com/docker/distribution/digest"
 	storagedriver "github.com/docker/distribution/registry/storage/driver"
-	"github.com/docker/docker/pkg/tarsum"
```
Comment (diogomonica) [inline registry/storage/layerupload.go]: This makes me so happy.

**Comment by jlhawn** [comment]
LGTM

**Comment by jlhawn** [comment]
@stevvooe who do we need to review this? @dmp42 ?

File: registry/handlers/api_test.go [hunk 3]
```diff
@@ -520,7 +520,11 @@ func doPushLayer(t *testing.T, ub *v2.URLBuilder, name string, dgst digest.Diges
 
 // pushLayer pushes the layer content returning the url on success.
 func pushLayer(t *testing.T, ub *v2.URLBuilder, name string, dgst digest.Digest, uploadURLBase string, body io.Reader) string {
-	resp, err := doPushLayer(t, ub, name, dgst, uploadURLBase, body)
+	buf := new(bytes.Buffer)
```
Comment (stevvooe) [inline registry/handlers/api_test.go]: Why is this necessary?

Comment (stevvooe) [inline registry/handlers/api_test.go]: I see now: body is needed for `doPushLayer` and `digest.FromBytes`. Rather than copy the whole, potentially large buffer, you can use an `io.TeeReader`:

```
dgster = digest.NewDigesterDefault()
resp, err := doPushLayer(t, ub, name, dgst, uploadURLBase,  io.TeeReader(body, dgster))
...
dgst := dgster.Digest()
```

We'll have to add this to the digest package:

```
// Digester calculates the digest of written data. It is functionally
// equivalent to hash.Hash but provides methods for returning the Digest type
// rather than raw bytes.
type Digester struct {
	alg  string
	hash hash.Hash
}

func NewDigester(alg string, h hash.Hash) Digester {
	return Digester{
		alg:  alg,
		hash: h,
	}
}

func NewDigesterDefault() Digester {
	return NewDigester("sha256", sha256.New())
}

// Write data to the digester. These writes cannot fail.
func (d *Digester) Write(p []byte) (n int, err error) {
	return d.hash.Write(p)
}

// Digest returns the current digest for this digester.
func (d *Digester) Digest() Digest {
	return NewDigest(d.alg, d.hash)
}

// Reset the state of the digester.
func (d *Digester) Reset() {
	d.hash.Reset()
}
```

**Comment by stevvooe** [comment]
LGTM except for `Digester` addition (I know it's just tests but the allocations can slow things down).

File: registry/handlers/api_test.go (hunk 3 above)
Comment (dmcgowan) [inline registry/handlers/api_test.go]: +1 for using `io.TeeReader`, its the approach been using on engine code and its pretty siimple

Comment (jlhawn) [inline registry/handlers/api_test.go]: We can do this now only because we're not supporting tarsum anymore. (Tarsum was an `io.Reader` not an `io.Writer`) :dancer:

Comment (stevvooe) [inline registry/handlers/api_test.go]: :japanese_ogre:

Comment (endophage) [inline registry/handlers/api_test.go]: I could also just change doPushLayer to accept a body of []byte rather than an io.Reader and it would be a much smaller change

@stevvooe would that work for you?

File: digest/digester.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+// Digester calculates the digest of written data. It is functionally
+// equivalent to hash.Hash but provides methods for returning the Digest type
+// rather than raw bytes.
+type Digester struct {
+	alg  string
+	hash hash.Hash
+}
+
+func NewDigester(alg string, h hash.Hash) Digester {
```
Comment (dmcgowan) [inline digest/digester.go]: Comment needed on this function and the next to pass lint

**Comment by dmcgowan** [comment]
LGTM after lint fix

**Comment by stevvooe** [comment]
LGTM!

Please fix function documentation per @dmcgowan.
